cmsl1t/io/eventreader.py

```diff
@@ -0,0 +1,135 @@
+    input_files = []
+    for p in paths:
+        if '*' in p:
+            input_files.extend(glob(p))
+        else:
+            input_files.append(p)
+    return input_files
+
+
+def _create_alias_map(ntuple_map):
+    '''
+        Translates the ntuple_map into a format that can more easily processed
+        for creating event attributes, e.g.
+        content:
+            <tree path>:
+                branches:
+                  <branchname>.<leafname>:
+                    aliases:
+                      - event.myAlias
+        to
+        {'myAlias': (<tree path>, <branchname>.<leafname>)}
+
+    '''
+    # TODO: move this replacement into cmsl1t.config
+    aliasMap = {}
+    trees = ntuple_map['content']
+    for treeName, content in trees.items():
+        for branchName, branchContent in content['branches'].items():
+            for alias in branchContent['aliases']:
+                tidyAlias = alias.replace('event.', '')
+                aliasMap[tidyAlias] = (treeName, branchName)
+    return aliasMap
+
+
+class EventReader(object):
+
+    def __init__(self, input_files, ntuple_map, nevents=-1):
+        '''
+            Reads ntuple_info as defined by bin/create-map-file
+        '''
+        self._treeNames = ntuple_map['content'].keys()
+        self._aliasMap = _create_alias_map(ntuple_map)
+        self.input_files = _get_input_files(input_files)
+        self.nevents = nevents
+        self._trees = {}
+        for alias in self._aliasMap:
+            if 'vertex' in alias.lower():
+                print(alias)
+
+        self._load_trees()
+
+    def _load_trees(self):
+        for treeName in self._treeNames:
+            try:
+                self._trees[treeName] = TreeChain(
+                    treeName,
+                    self.input_files,
+                    cache=True,
+                    events=self.nevents,
+                )
+            except RuntimeError:
+                logger.warn(
+                    "Cannot find tree: {0} in input file".format(treeName))
+                continue
+
+    def contains(self, name):
+        return name in self._aliasMap.keys()
+
+    def __iter__(self):
+        # event loop
+        for trees in six.moves.zip(*self._trees.itervalues()):
+            yield Event(self._trees, self._aliasMap)
+
+
+class Event(object):
+
+    energySumTypes = {
+        'Ett': sumTypes.kTotalEt,
+        'EttHF': sumTypes.kTotalEtHF,
+        'Htt': sumTypes.kTotalHt,
+        'HttHF': sumTypes.kTotalHtHF,
+        'Met': sumTypes.kMissingEt,
+        'MetHF': sumTypes.kMissingEtHF,
+        'Mht': sumTypes.kMissingHt,
+        'Mex': sumTypes.kTotalEtx,
+        'Mey': sumTypes.kTotalEty,
+    }
+
+    def __init__(self, trees, mapping):
+        self._map = mapping
+        self._trees = trees
+        self._cache = {}
+
+    def __getattr__(self, name):
+        if name in object.__getattribute__(self, '_cache'):
+            return object.__getattribute__(self, '_cache')[name]
```

Comment:
For these sorts of things, and since this is code that will run often from within the event loop, I think it's better to do:
```
thing = object.__getattribute__(self, '_cache').get(name, impossible_to_find_in_tree_object)
if thing is not impossible_to_find_in_tree_object:
    return thing
```
This would only need a single look-up inside the `_cache`.   For `impossible_to_find_in_tree_object`, I'd just make a private class within object this file that only the Event class would know about.  

You could use `None` instead of some special object like `impossible_to_find_in_tree_object`:
```
thing = object.__getattribute__(self, '_cache').get(name, None)
if thing:
    return thing
```
but perhaps the risk here is that there really is a NULL pointer stored in the tree

Reply:
I would prefer `None`. In the (very) rare cases that you have a null-pointer, it will simply ignore the cache.